modules/cluster-management/src/test/java/org/apache/ignite/internal/cluster/management/LogicalTopologyJoinAwaiterTest.java

```diff
@@ -0,0 +1,155 @@
+import org.apache.ignite.internal.cluster.management.topology.api.LogicalNode;
+import org.apache.ignite.internal.network.ClusterNodeImpl;
+import org.apache.ignite.internal.testframework.matchers.CompletableFutureMatcher;
+import org.apache.ignite.network.ClusterNode;
+import org.apache.ignite.network.NetworkAddress;
+import org.junit.jupiter.api.Timeout;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+class LogicalTopologyJoinAwaiterTest {
+    private static final ClusterNode node1 = newClusterNode();
+    private static final ClusterNode node2 = newClusterNode();
+    private static final ClusterNode node3 = newClusterNode();
+    private static final ClusterNode node4 = newClusterNode();
+
+    @ParameterizedTest(name = ParameterizedTest.INDEX_PLACEHOLDER + " positiveLogicalTopologyJoinTest: {0}")
+    @Timeout(5)
+    @MethodSource("positiveClusterBehaviour")
+    void positiveLogicalTopologyJoinTest(
+            BiConsumer<PhysicalTopologySimulator, LogicalTopologySimulator> behaviourSimulation
+    ) throws Exception {
+        Set<ClusterNode> initialPhysicalTopology = Set.of(node1, node2, node3, node4);
+        PhysicalTopologySimulator physicalTopologySimulator = new PhysicalTopologySimulator(initialPhysicalTopology);
+        LogicalTopologySimulator logicalTopologySimulator = new LogicalTopologySimulator(emptySet());
+
+        ExecutorService simulationExecutor = Executors.newSingleThreadExecutor();
+        try {
+            CompletableFuture<Void> allNodesJoinedLogicalTopology =
+                    awaitLogicalTopologyToMatchPhysicalTopology(physicalTopologySimulator, logicalTopologySimulator);
+
+            Future<?> simulationJob = simulationExecutor.submit(() -> {
+                behaviourSimulation.accept(physicalTopologySimulator, logicalTopologySimulator);
+            });
+
+            simulationJob.get(5, TimeUnit.SECONDS);
+
+            assertThat(allNodesJoinedLogicalTopology, CompletableFutureMatcher.willCompleteSuccessfully());
+            assertTrue(physicalTopologySimulator.hasNoEventHandlers());
+            assertTrue(logicalTopologySimulator.hasNoEventHandlers());
+        } finally {
+            assertThat(simulationExecutor.shutdownNow(), emptyIterable());
+        }
+    }
+
+    private static Stream<BiConsumer<PhysicalTopologySimulator, LogicalTopologySimulator>> positiveClusterBehaviour() {
+        return Stream.of(
+                simulation("all nodes join logical topology normally", (physicalTopology, logicalTopology) -> {
+                    logicalTopology.putNode(logical(node1));
+                    logicalTopology.putNode(logical(node2));
+                    logicalTopology.putNode(logical(node3));
+                    logicalTopology.putNode(logical(node4));
+                }),
+                simulation("3 nodes join logical topology and node4 leaves physical topology", (physicalTopology, logicalTopology) -> {
```

Comment:
This simulation makes logical topology to be equal to the physical eventually. You assert the await future is done eventually. But you do not test the future state during the topology changes. I would like to see asserts on the future after each topology node, for example: 

```java
logicalTopology.putNode(logical(node1));
assertFutureIsNotCompleted();
logicalTopology.putNode(logical(node2));
assertFutureIsNotCompleted();
logicalTopology.putNode(logical(node3));
assertFutureIsNotCompleted();
logicalTopology.putNode(logical(node4));
assertFutureCompleted();